File: backend/src/llm.ts
import OpenAI from 'openai';
import { Node } from './node';
const openai = new OpenAI();

const generateTitle = async (text: string): Promise<string> => {
  const result = await openai.chat.completions.create({
    model: 'gpt-3.5-turbo',
    messages: [
      {
        role: 'user',
        content: `Given the following content, which may have both text and web page URLs, please provide a concise and descriptive title: /n ${text}`,
      },
    ],
    functions: [
      {
        name: 'generateContentTitle',
        description:
          'Given some text content and URLs, create a short concise title.',
        parameters: {
          type: 'object',
          properties: {
            title: {
              type: 'string',
              description: 'The title of the content.',
            },
          },
          required: ['title'],
        },
      },
    ],
    function_call: { name: 'generateContentTitle' },
  });

  const funcArgs = result.choices[0].message.function_call?.arguments;
  if (!funcArgs) {
    throw new Error(
      `Failed to generate title (no function arguments) for node text: ${text}`,
    );
  }
  const title = JSON.parse(funcArgs).title;
  if (!title) {
    throw new Error(`Failed to generate title for node text: ${text}`);
  }
  return title;
};

const summariseOrAnswerFromDocuments = async (nodes: Node[], query: string) => {
  const titles = nodes.map((node) => `- ${node.toDTO().title}`).join('\n');
  const input = `
  Given the following list of documents, and query, you must
  answer the question, or summarise the docs depending on the tone.
  ----
  DOCUMENTS:
  ${titles}
  ----
  QUERY:
  ${query}
  `;
  const stream = await openai.chat.completions.create({
    model: 'gpt-3.5-turbo',
    messages: [{ role: 'user', content: input }],
    stream: true,
  });
  return stream;
};

export { generateTitle, summariseOrAnswerFromDocuments };

File: backend/src/llm.py
import json
from typing import List
from openai import OpenAI

client = OpenAI()

MODEL = "gpt-3.5-turbo-0613"

PROMPT_TEMPLATE = (
    "A question and context documents are provided below."
    "If the required information is not available in the context documents to answer the question,"
    "explain to the user that you are missing the required information."
    "If you are able to answer make sure you quote the source."
    "You will be provided a JSON line for each context document, which contains <text>, <title> & <url> keys."
    "At the END of your reply (not inline), you should list all unique sources using the <title> & <url> keys."
    "Answer using Markdown."
    "---------------------\n"
    "CONTEXT:\n"
    "{context}"
    "---------------------\n"
    "QUESTION:\n"
    "{question}"
)


def format_chunks(chunks: List[dict]) -> str:
    result = ""
    for chunk in chunks:
        chunk.pop("id")
        chunk.pop("score")
        result += f"{json.dumps(chunk)}\n"
    return result


def answer_with_context(chunks: List[dict], question: str) -> str:
    formatted_chunks = format_chunks(chunks)
    messages = [
        {
            "role": "user",
            "content": PROMPT_TEMPLATE.format(
                context=formatted_chunks, question=question
            ),
        }
    ]
    stream = client.chat.completions.create(
        messages=messages,
        model=MODEL,
        stream=True,
    )
    for chunk in stream:
        content = chunk.choices[0].delta.content
        if content:
            yield content
        else:
            yield ""
        # yield chunk.choices[0].delta.content
    # return chat_completion.choices[0].message.content
